File: 1102.cpp
#include <bits/stdc++.h>
constexpr int inf = 987654321;
using namespace std;

int cost[17][17];
bool on[17];
int dp[100000];
int P;
int N;
int state;

int dfs(int cnt) {
    if(cnt == P) {
        return 0;
    }
    
    int &cur_cost = dp[state];
    if(cur_cost != -1)
        return cur_cost;
    
    cur_cost = inf;
    
    for(int i = 0; i < N; i++) {
        // i가 안켜져있으면
        if (!((1 << i) & state)) {
            for(int j = 0; j < N; j++) {
                if(i == j)
                    continue;
                // j가 켜져있으면
                if((1 << j) & state) {
                    state |= (1 << i);
                    cur_cost = min(cur_cost, dfs(cnt + 1) + cost[j][i]);
                    state ^= (1 << i);
                }
            }
        }
    }
    
    return cur_cost;
}

int main() {
    ios_base :: sync_with_stdio(false); 
    cin.tie(NULL); 
    cout.tie(NULL);
    memset(dp, -1, sizeof(dp));
    
    cin >> N;
    
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < N; j++) {
            cin >> cost[i][j];
        }
    }
    
    string str;
    
    cin >> str;
    cin >> P;
    
    for(int i = 0; i < N; i++) {
        if(str[i] == 'Y') {
            P--;
            state |= (1 << i);
        }
    }
    
    int ans = dfs(0);
    
    if (P <= 0)
        cout << 0;
    else if (ans == inf) 
        cout << -1;
    else
        cout << ans;
}

File: 1102.py
import sys
input = sys.stdin.readline

N = int(input())

dp = [-1 for i in range(100000)]

cost = [list(map(int, input().split())) for i in range(N)]
on = list(input()[:-1])
P = int(input())

def dfs(state, cnt):
    global cost, ans
    
    if cnt == P:
        return 0
    cur_cost = dp[state]  
    
    if cur_cost != -1:
        return cur_cost
    
    cur_cost = float('inf')
    
    for i in range(N):
        # i가 안켜져있으면
        if not ((1 << i) & state):
            for j in range(N):
                if i == j:
                    continue
                
                if ((1 << j) & state):
                    cur_cost = min(cur_cost, dfs(state | (1 << i), cnt + 1) + cost[j][i])
    return cur_cost

state = 0
for i in range(N):
    if on[i] == 'Y':
        P -= 1
        state |= (1 << i)
        
ans = dfs(state, 0)

if (P <= 0):
    print(0)
elif ans == float('inf'):
    print(-1)
else:
    print(ans)